File: src/main/scala/mr/merc/sound/JavaLayerMp3Renderer.java
package mr.merc.sound;

//JFC
import java.io.BufferedInputStream;
import java.io.IOException;
import java.net.URL;

import javazoom.jl.decoder.JavaLayerException;
import javazoom.jl.player.FactoryRegistry;
import javazoom.jl.player.Player;
import javazoom.jl.player.advanced.AdvancedPlayer;
import javazoom.jl.player.advanced.PlaybackEvent;
import javazoom.jl.player.advanced.PlaybackListener;

// TODO rewrite into scala
public class JavaLayerMp3Renderer extends BaseAudioRenderer {

	private AdvancedPlayer player;

	@Override
	protected void playSound(final URL audiofile) {
		try {
			player = new AdvancedPlayer(new BufferedInputStream(
					audiofile.openStream()), FactoryRegistry.systemRegistry()
					.createAudioDevice());

			player.setPlayBackListener(new PlaybackListener() {
				@Override
				public void playbackFinished(PlaybackEvent pe) {
					super.playbackFinished(pe);
					setStatus(Status.END_OF_SOUND);
				}
			});
			player.play();
		} catch (IOException e) {
			setStatus(Status.ERROR);
			System.err.println("Can not load audiofile (" + audiofile + ": "
					+ e);
		} catch (JavaLayerException e) {
			setStatus(Status.ERROR);
			System.err.println("Problem playing audio: " + e);
		}
	}
}

File: src/main/scala/mr/merc/sound/JavaLayerMp3Renderer.scala
package mr.merc.sound

import java.net.URL
import javazoom.jl.player.advanced.AdvancedPlayer
import java.io.BufferedInputStream
import javazoom.jl.player.FactoryRegistry
import javazoom.jl.player.advanced.PlaybackListener
import javazoom.jl.player.advanced.PlaybackEvent
import java.io.InputStream

class JavaLayerMp3Renderer(listener: Status => Unit) extends BaseAudioRenderer(listener) {

  def playSound(audiofile: InputStream) {
    val player = new AdvancedPlayer(new BufferedInputStream(audiofile),
      FactoryRegistry.systemRegistry().createAudioDevice())

    player.setPlayBackListener(new PlaybackListener() {
      override def playbackFinished(pe: PlaybackEvent) {
        super.playbackFinished(pe)
        status = EndOfSound
      }
    })
    player.play()
  }
}